Kafka, Schema Registry and Avro guide (#16503)
quarkusio/quarkus: docs/src/main/asciidoc/kafka-schema-registry-avro.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
////
= Quarkus - Using Apache Kafka with Schema Registry and Avro

include::./attributes.adoc[]

This guide demonstrates how your Quarkus application can use Schema Registry with Avro.

The guide is based on the link:https://quarkus.io/blog/kafka-avro/[How to Use Kafka, Schema Registry and Avro with Quarkus] blog post.

> If you are not familiar with Avro or the concept of Schema Registry, the post has a good introduction on the subject.

Author: hmm... @cescoffier WDYT?

[source,bash,subs=attributes+]
----
mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=kafka-avro-schema-quickstart \
    -DclassName="org.acme.kafka.MovieResource" \
    -Dpath="/movies" \
    -Dextensions="resteasy-reactive,resteasy-reactive-jackson,rest-client,smallrye-reactive-messaging-kafka,avro"
cd kafka-avro-schema-quickstart
----

Please note that even thought our application will not explicitly use a REST client, it is used
under the hood by the Apicurio registry and, at the moment it is required as a dependency.

Additionally, we need a serializer and a deserializer for Avro, specific to the schema registry we are going to use.
Apicurio, which we use in this guide, has them in the following dependency:

[source,xml]
----
<dependency>
    <groupId>io.apicurio</groupId>
    <artifactId>apicurio-registry-utils-serde</artifactId>
    <version>1.2.2.Final</version>
    <exclusions>
        <exclusion>
            <groupId>org.jboss.spec.javax.interceptor</groupId>
            <artifactId>jboss-interceptors-api_1.2_spec</artifactId>
        </exclusion>
    </exclusions>
</dependency>
----

== Avro schema

> Let's create a file called `src/main/avro/movie.avsc` with the schema for our record (Kafka message):

Reviewer: ```suggestion
Apache Avro is a data serialization system. Data structured are described using schemas. 
So the first thing to do is creating a schema describing the `Movie` structure.
Create a file called `src/main/avro/movie.avsc` with the schema for our record (Kafka message):
```

[source,javascript]
----
{
  "namespace": "org.acme.kafka.quarkus",
  "type": "record",
  "name": "Movie",
  "fields": [
    {
      "name": "title",
      "type": "string"
    },
    {
      "name": "year",
      "type": "int"
    }
  ]
}
----

If you build the project with `mvn compile` , the `movies.avsc` will get compiled to a `Movie.java` file

> placed in the `target/generated-sources/avsc` directory.

Reviewer: ```suggestion
placed in the `target/generated-sources/avsc` directory.

TIP: With Quarkus, no need to a specific plugin to process the Avro schema, this is all done for you!
```

> placed in the `target/generated-sources/avsc` directory.

Author: automatic

If you run the project with `mvn compile quarkus:dev`, the changes you do to the schema file will be
automatically applied to the generated Java files.

== The `Movie` producer
Having defined the record, we can now jump to implementing the `MovieResource`.

> Let's open the `MovieResource`, inject an `Emitter` of `Movie` DTO and implement a `@POST` method

Reviewer: ```suggestion
Let's open the `MovieResource`, inject an https://quarkus.io/blog/reactive-messaging-emitter/[`Emitter`] of `Movie` DTO and implement a `@POST` method
```

that consumes `Movie` and sends it through the `Emitter`:

[source,java]
----
package org.acme.kafka;

import org.acme.kafka.quarkus.Movie;
import org.eclipse.microprofile.reactive.messaging.Channel;
import org.eclipse.microprofile.reactive.messaging.Emitter;
import org.jboss.logging.Logger;

import javax.ws.rs.Consumes;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

@Path("/movies")
@Consumes(MediaType.APPLICATION_JSON)

> @Produces(MediaType.APPLICATION_JSON)

Reviewer: @geoand Isn't JSON the default in RESTEasy Reactive? Wondering if we can remove the 2 lines.

Reviewer 2: Yes, exactly. They can be removed

Author: 👍🏻

public class MovieResource {

    private static final Logger LOGGER = Logger.getLogger(MovieResource.class);

>     @Channel("movies") Emitter<Movie> emitter;

Reviewer: Isn't `@Inject` missing? Maybe it's not needed and I never realized it.

Reviewer 3: This is a non-standard extension to CDI in Quarkus. If you use a qualifier, you don't have to add `@Inject`.

    @POST
    public Response enqueueMovie(Movie movie) {
        LOGGER.infof("Sending movie %s to Kafka",
                movie.getTitle()
        );
        emitter.send(movie);
        return Response.accepted().build();
    }

}
----

> Now, configure the channel for the `Emitter` in `application.properties`:

Reviewer: ```suggestion
Now, we need to _map_ the `movies` channel (the `Emitter` emits to this channel) to a Kafka topic. 
To achieve this, edit the `application.properties` file, and add the following content:
```

[source,properties]
----
# set the URL of the Apicurio Schema Registry, a global setting shared between producers and consumers
mp.messaging.connector.smallrye-kafka.apicurio.registry.url=http://localhost:8081/api

# set the connector to use for the `movies` channel to smallrye-kafka
mp.messaging.outgoing.movies.connector=smallrye-kafka

# the name of the corresponding Kafka topic to `movies`
mp.messaging.outgoing.movies.topic=movies

# set the serializer for the `movies` channel to the Avro Serializer for Apicurio
mp.messaging.outgoing.movies.value.serializer=io.apicurio.registry.utils.serde.AvroKafkaSerializer

# Apicurio schema specific settings:
mp.messaging.outgoing.movies.apicurio.registry.artifact-id=io.apicurio.registry.utils.serde.strategy.SimpleTopicIdStrategy
mp.messaging.outgoing.movies.apicurio.registry.global-id=io.apicurio.registry.utils.serde.strategy.GetOrCreateIdStrategy
mp.messaging.outgoing.movies.apicurio.registry.avro-datum-provider=io.apicurio.registry.utils.serde.avro.ReflectAvroDatumProvider
----

== The `Movie` consumer

> We already have means to put `Movie` records to a Kafka topic, it's time to implement a consumer for them.

Reviewer: ```suggestion
So, we can write records into Kafka containing our `Movie` data.
That data is serialized using Avro. 
Now, it's time to implement a consumer for them.
```

Let's create `ConsumedMovieResource` that will consume `Movie` messages

> from the `movies-from-kafka` channel and will expose it via Server Sent Events:

Reviewer: ```suggestion
from the `movies-from-kafka` channel and will expose it via Server-Sent-Events:
```

Reviewer 3: The canonical name seems to be "Server-sent events" (or "Server-Sent Events" if you want the words capitalized).

I looked at:

- https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events
- https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events
- https://en.wikipedia.org/wiki/Server-sent_events

[source,java]
----
package org.acme.kafka;

import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

import org.acme.kafka.quarkus.Movie;
import org.eclipse.microprofile.reactive.messaging.Channel;
import org.jboss.resteasy.annotations.SseElementType;

import io.smallrye.mutiny.Multi;

@ApplicationScoped
@Path("/consumed-movies")
public class ConsumedMovieResource {

>     @Inject

Reviewer: So, if we didn't need the `@Inject` before, why do we need it here?

    @Channel("movies-from-kafka")

>     Multi<Movie> prices;

Author: lol

    @GET
    @Produces(MediaType.SERVER_SENT_EVENTS)

>     @SseElementType("text/plain")

Reviewer: Use MediaType instead

    public Multi<String> stream() {

>         return prices.map(movie -> String.format("'%s' from %s", movie.getTitle(), movie.getYear()));

Author: ```suggestion
        return movies.map(movie -> String.format("'%s' from %s", movie.getTitle(), movie.getYear()));
```

    }
}
----

> The last bit of the application's code is the configurations of the `movies-from-kafka` channel in

Reviewer: ```suggestion
The last bit of the application's code is the configuration of the `movies-from-kafka` channel in
```

mp.messaging.incoming.movies-from-kafka.topic=movies

# set the deserializer for the `movies-from-kafka` channel to the Avro Deserializer for Apicurio
mp.messaging.incoming.movies-from-kafka.value.deserializer=io.apicurio.registry.utils.serde.AvroKafkaDeserializer

# disable auto-commit, Reactive Messaging handles it itself
mp.messaging.incoming.movies-from-kafka.enable.auto.commit=false

mp.messaging.incoming.movies-from-kafka.auto.offset.reset=earliest
mp.messaging.incoming.movies-from-kafka.apicurio.registry.avro-datum-provider=io.apicurio.registry.utils.serde.avro.ReflectAvroDatumProvider
----

== The infrastructure
To use our application, we need Kafka and Apicurio Schema Registry.
The easiest way to get them running is to use `docker-compose` to start the appropriate containers.

Create a `docker-compose.yaml` file at the root of the project with the following content:

[source,yaml]
----
version: '2'

services:

  zookeeper:
    image: strimzi/kafka:0.11.3-kafka-2.1.0
    command: [
      "sh", "-c",
      "bin/zookeeper-server-start.sh config/zookeeper.properties"
    ]
    ports:
      - "2181:2181"
    environment:
      LOG_DIR: /tmp/logs

  kafka:

>     image: strimzi/kafka:0.11.3-kafka-2.1.0

Reviewer: That sounds quite out of date.

Author: updated, and it still works :)

        <artifactId>testcontainers</artifactId>
        <scope>test</scope>
    </dependency>
    <dependency>
        <groupId>io.strimzi</groupId>
        <artifactId>strimzi-test-container</artifactId>
        <version>0.22.1</version>
        <scope>test</scope>
        <exclusions>
            <exclusion>
                <groupId>org.apache.logging.log4j</groupId>
                <artifactId>log4j-core</artifactId>
            </exclusion>
        </exclusions>
    </dependency>
    <dependency>
        <groupId>org.awaitility</groupId>
        <artifactId>awaitility</artifactId>
        <scope>test</scope>
    </dependency>
</dependencies>
----

Now, let's define a link:getting-started-testing.adoc#quarkus-test-resource[QuarkusTestResourceLifecycleManager] that will create the appropriate
containers:
[source,java]
----
package org.acme.kafka;

import java.util.HashMap;
import java.util.Map;

import org.testcontainers.containers.GenericContainer;

import io.quarkus.test.common.QuarkusTestResourceLifecycleManager;
import io.strimzi.StrimziKafkaContainer;

public class KafkaAndSchemaRegistryTestResource implements QuarkusTestResourceLifecycleManager {

>     private static final StrimziKafkaContainer kafka = new StrimziKafkaContainer();

Reviewer 3: Does it need to be `static`?

Author: works without it, I'll drop `static`

>     private static GenericContainer<?> registry;

Reviewer 3: Does it need to be `static`?

    @Override
    public Map<String, String> start() {
        kafka.start();
        registry = new GenericContainer<>("apicurio/apicurio-registry-mem:1.2.2.Final")
                .withExposedPorts(8080)
                .withEnv("QUARKUS_PROFILE", "prod")
                .withEnv("KAFKA_BOOTSTRAP_SERVERS", kafka.getBootstrapServers())
                .withEnv("APPLICATION_ID", "registry_id")

>                 .withEnv("APPLICATION_SERVER", "localhost:9000");

Reviewer 3: Could this instantiation be also inlined in the `registry` field declaration, just like `kafka = new StrimziKafkaContainer`?

Author: kafka.getBootstrapServers() requires the mapped port for Kafka, so I don't think it can be used before kafka.start()

Reviewer 3: Ah sorry, didn't notice that one!

        registry.start();
        Map<String, String> properties = new HashMap<>();
        properties.put("mp.messaging.connector.smallrye-kafka.apicurio.registry.url",
                "http://" + registry.getContainerIpAddress() + ":" + registry.getMappedPort(8080) + "/api");
        properties.put("kafka.bootstrap.servers", kafka.getBootstrapServers());
        return properties;
    }

    @Override
    public void stop() {
        registry.stop();

>         kafka.close();

Reviewer: What's the difference between `stop` and `close`? I know I wrote that code, that does not mean I know what I'm doing :-D

Author: `close` calls `stop` :D I will change to `stop` for consistency

                                        .contentType(ContentType.JSON)
                                        .body("{\"title\":\"The Shawshank Redemption\",\"year\":1994}")
                                    .when().post("/movies")
                                    .then()
                                        .statusCode(202);
                                given()
                                        .contentType(ContentType.JSON)
                                        .body("{\"title\":\"12 Angry Men\",\"year\":1957}")
                                    .when().post("/movies")
                                    .then()
                                        .statusCode(202);
                                try {
                                    Thread.sleep(200L);
                                } catch (InterruptedException e) {
                                    break;
                                }
                            }
                        }
                );
        return executorService;
    }

}
----

== Avro code generation details
In this guide we used the Quarkus code generation mechanism to generate Java files

> from Avro schema.

Reviewer 3: Ha, I knew it! :-)

> Under the hoods, the mechanism uses `org.apache.avro:avro-compiler`.

Reviewer: ```suggestion
Under the hood, the mechanism uses `org.apache.avro:avro-compiler`.
```

> Under the hoods, the mechanism uses `org.apache.avro:avro-compiler`

Reviewer 3: Missing `.` at the end.

Author: ```suggestion
Under the hoods, the mechanism uses `org.apache.avro:avro-compiler`.
```

> You can use the following properties to alter how it works:

Reviewer 3: Maybe

```suggestion
You can use the following configuration properties to alter how it works:
```

?

Author: why not

- `avro.codegen.[avsc|avdl|avpr].imports` - a list of files or directories that should be compiled first thus making them

> importable by subsequently compiled schemas. Note that imported files should not reference each other. All paths should be relative to the src/[main|test]/avro directory. Passed as a comma-separated list.

Reviewer: ```suggestion
importable by subsequently compiled schemas. Note that imported files should not reference each other. All paths should be relative to the `src/[main|test]/avro` directory. Passed as a comma-separated list.
```

- `avro.codegen.stringType` - the Java type to use for Avro strings. May be one of `CharSequence`, `String` or
`Utf8`. Defaults to `String`
- `avro.codegen.createOptionalGetters` - enables generating the `getOptional...`
methods that return an Optional of the requested type. Defaults to `false`
- `avro.codegen.enableDecimalLogicalType` - determines whether to use Java classes for decimal types, defaults to `false`
- `avro.codegen.createSetters` - determines whether to create setters for the fields of the record.
Defaults to `false`
- `avro.codegen.gettersReturnOptional` - enables generating `get...` methods that
return an Optional of the requested type. Defaults to `false`
- `avro.codegen.optionalGettersForNullableFieldsOnly`, works in conjunction with `gettersReturnOptional` option.
If it is set, `Optional` getters will be generated only for fields that are nullable. If the field is mandatory,
regular getter will be generated. Defaults to `false`

== Further reading
This guide used SmallRye Reactive Messaging Kafka to produce Kafka messages and consume messages from Kafka, to learn more about

> this project, take a look at link:https://smallrye.io/smallrye-reactive-messaging/smallrye-reactive-messaging/2/kafka/kafka.html[its guide].

Reviewer: Add a link to the original blog post here.

Author: done

records, and connect to Schema Registry (such as the https://docs.confluent.io/platform/current/schema-registry/index.html[Confluent Schema Registry] or https://www.apicur.io/registry/[Apicurio Schema Registry].

If you are not familiar with Kafka and Kafka in Quarkus in particular, consider
first going through the link:kafka.adoc[Using Apache Kafka with Reactive Messaging] guide.

== Prerequisites

To complete this guide, you need:

* less than 30 minutes
* an IDE
* JDK 11+ installed with `JAVA_HOME` configured appropriately
* Apache Maven {maven-version}
* Docker Compose to start a local Kafka cluster and Apicurio Registry
* GraalVM installed if you want to run in native mode.

== Architecture

In this guide we are going to implement a REST resource, namely `MovieResource` that
will consume movie DTOs and put them in a Kafka topic.

Then, we will implement a consumer that will consume and collect messages from the same topic.
The collected messages will be then exposed by another resource, `ConsumedMovieResource`, via
https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events[Server-Sent Events].

> 

Reviewer: ```suggestion
The _Movies_ will be serialized and deserialized using Avro. 
The schema, describing the _Movie_, is stored in an Apicurio schema registry.
The same concept applies if you are using the Confluent Avro _serde_ and Confluent Schema Registry.
```

                <artifactId>jakarta.xml.bind-api</artifactId>
            </exclusion>
            <exclusion>
                <groupId>jakarta.ws.rs</groupId>
                <artifactId>jakarta.ws.rs-api</artifactId>
            </exclusion>
        </exclusions>
    </dependency>
</dependencies>
<repositories>
    <repository>
        <id>confluent</id>
        <url>https://packages.confluent.io/maven/</url>
        <snapshots>
            <enabled>false</enabled>
        </snapshots>
    </repository>
</repositories>
----
====

== Avro schema
Apache Avro is a data serialization system. Data structures are described using schemas.
The first thing we need to do is to create a schema describing the `Movie` structure.
Create a file called `src/main/avro/movie.avsc` with the schema for our record (Kafka message):
[source,javascript]
----
{
  "namespace": "org.acme.kafka.quarkus",
  "type": "record",
  "name": "Movie",
  "fields": [
    {
      "name": "title",
      "type": "string"
    },
    {
      "name": "year",
      "type": "int"
    }
  ]
}
----

If you build the project with `mvn compile` , the `movies.avsc` will get compiled to a `Movie.java` file

> placed in the `target/generated-sources/avsc` directory.

Reviewer: ```suggestion
placed in the `target/generated-sources/avsc` directory. Refer to the https://avro.apache.org/docs/current/spec.html#schemas[Avro specification] to learn more about the Avro syntax and supported types.
```

Author: 👍🏻

TIP: With Quarkus, no need to a specific plugin to process the Avro schema, this is all done for you!

If you run the project with `mvn compile quarkus:dev`, the changes you do to the schema file will be
automatically applied to the generated Java files.

== The `Movie` producer

> Having defined the record, we can now jump to implementing the `MovieResource`.

Reviewer: ```suggestion
Having defined the schema, we can now jump to implementing the `MovieResource`.
```

    public Multi<String> stream() {
        return movies.map(movie -> String.format("'%s' from %s", movie.getTitle(), movie.getYear()));
    }
}
----

The last bit of the application's code is the configuration of the `movies-from-kafka` channel in
`application.properties`:
[source,properties]
----
# set the connector for the incoming channel to `smallrye-kafka`
mp.messaging.incoming.movies-from-kafka.connector=smallrye-kafka

# set the topic name for the channel to `movies`
mp.messaging.incoming.movies-from-kafka.topic=movies

# set the deserializer for the `movies-from-kafka` channel to the Avro Deserializer for Apicurio
mp.messaging.incoming.movies-from-kafka.value.deserializer=io.apicurio.registry.utils.serde.AvroKafkaDeserializer

# disable auto-commit, Reactive Messaging handles it itself
mp.messaging.incoming.movies-from-kafka.enable.auto.commit=false

mp.messaging.incoming.movies-from-kafka.auto.offset.reset=earliest
mp.messaging.incoming.movies-from-kafka.apicurio.registry.avro-datum-provider=io.apicurio.registry.utils.serde.avro.ReflectAvroDatumProvider
----

== The infrastructure
To use our application, we need Kafka and Apicurio Schema Registry.

> The easiest way to get them running is to use `docker-compose` to start the appropriate containers.

Reviewer: Here is a tricky part... `docker-compose` became `docker compose` in the last docker version. I would keep it like this for now, but we may have to update (globally) the reference.

[source,shell script]
----
curl -N http://localhost:8080/consumed-movies
----

In the third one, post a few movies:
[source,shell script]
----
curl --header "Content-Type: application/json" \
  --request POST \
  --data '{"title":"The Shawshank Redemption","year":1994}' \
  http://localhost:8080/movies

curl --header "Content-Type: application/json" \
  --request POST \
  --data '{"title":"The Godfather","year":1972}' \
  http://localhost:8080/movies

curl --header "Content-Type: application/json" \
  --request POST \
  --data '{"title":"The Dark Knight","year":2008}' \
  http://localhost:8080/movies

curl --header "Content-Type: application/json" \
  --request POST \
  --data '{"title":"12 Angry Men","year":1957}' \
  http://localhost:8080/movies
----

Observe what is printed in the second terminal. You should see something along the lines of:
[source]
----
data:'The Shawshank Redemption' from 1994

data:'The Godfather' from 1972

data:'The Dark Knight' from 2008

data:'12 Angry Men' from 1957
----

== Testing the application

=== Infrastructure for tests
We will now use Testcontainers to set up Kafka and Apicurio Schema Registry for tests.

> First, let's add test dependencies on Awaitility, Testcontainers and Strimzi to `pom.xml`:

Reviewer: I believe the Strimzi one is going to drag Test Containers.

Author: it does

                        () -> {
                            while (true) {
                                given()
                                        .contentType(ContentType.JSON)
                                        .body("{\"title\":\"The Shawshank Redemption\",\"year\":1994}")
                                    .when().post("/movies")
                                    .then()
                                        .statusCode(202);
                                given()
                                        .contentType(ContentType.JSON)
                                        .body("{\"title\":\"12 Angry Men\",\"year\":1957}")
                                    .when().post("/movies")
                                    .then()
                                        .statusCode(202);
                                try {
                                    Thread.sleep(200L);
                                } catch (InterruptedException e) {
                                    break;
                                }
                            }
                        }
                );
        return executorService;
    }

}
----

> 

Reviewer: Maybe add the native version? 
Avro should be supported in native.

Author: native version is already in the code generated by the quarkus plugin, I w ill mention it here and add info on how to execute it

importable by subsequently compiled schemas. Note that imported files should not reference each other. All paths should be relative
to the `src/[main|test]/avro` directory. Passed as a comma-separated list.
- `avro.codegen.stringType` - the Java type to use for Avro strings. May be one of `CharSequence`, `String` or
`Utf8`. Defaults to `String`
- `avro.codegen.createOptionalGetters` - enables generating the `getOptional...`
methods that return an Optional of the requested type. Defaults to `false`
- `avro.codegen.enableDecimalLogicalType` - determines whether to use Java classes for decimal types, defaults to `false`
- `avro.codegen.createSetters` - determines whether to create setters for the fields of the record.
Defaults to `false`
- `avro.codegen.gettersReturnOptional` - enables generating `get...` methods that
return an Optional of the requested type. Defaults to `false`
- `avro.codegen.optionalGettersForNullableFieldsOnly`, works in conjunction with `gettersReturnOptional` option.
If it is set, `Optional` getters will be generated only for fields that are nullable. If the field is mandatory,
regular getter will be generated. Defaults to `false`

== Further reading

> * link:https://smallrye.io/smallrye-reactive-messaging/smallrye-reactive-messaging/2/kafka/kafka.html[SmallRye Reactive Messaging Kafka] documentation

Reviewer: ```suggestion
* link:https://smallrye.io/smallrye-reactive-messaging/smallrye-reactive-messaging/2.9/kafka/kafka.html[SmallRye Reactive Messaging Kafka] documentation
```

That's a problem I have with Antora. It references a very old version (2 -> 2.0) while we are 2.9 and Avro support has been improved in 2.5.

                <artifactId>log4j-core</artifactId>
            </exclusion>
        </exclusions>
    </dependency>
    <dependency>
        <groupId>org.awaitility</groupId>
        <artifactId>awaitility</artifactId>
        <scope>test</scope>
    </dependency>
</dependencies>
----

Now, let's define a link:getting-started-testing.adoc#quarkus-test-resource[QuarkusTestResourceLifecycleManager] that will create the appropriate
containers:
[source,java]
----
package org.acme.kafka;

import java.util.HashMap;
import java.util.Map;

import org.testcontainers.containers.GenericContainer;

import io.quarkus.test.common.QuarkusTestResourceLifecycleManager;
import io.strimzi.StrimziKafkaContainer;

public class KafkaAndSchemaRegistryTestResource implements QuarkusTestResourceLifecycleManager {

    private final StrimziKafkaContainer kafka = new StrimziKafkaContainer();

    private GenericContainer<?> registry;

    @Override
    public Map<String, String> start() {
        kafka.start();
        registry = new GenericContainer<>("apicurio/apicurio-registry-mem:1.2.2.Final")
                .withExposedPorts(8080)
                .withEnv("QUARKUS_PROFILE", "prod")

>                 .withEnv("KAFKA_BOOTSTRAP_SERVERS", kafka.getBootstrapServers())

Reviewer 4: the env vars `KAFKA_BOOTSTRAP_SERVERS` , `APPLICATION_ID` and `APPLICATION_SERVER` are not required, the image `apicurio/apicurio-registry-mem` stores the schemas in-memory and do not require access to kafka

Reviewer 3: Good point. I'm updating the guide to use Apicurio Registry 2.0.0, so I'll remove this as part of the work.
